test/raft/unit/test_snapshot.c

```diff
@@ -284,3 +297,189 @@ TEST(snapshot, leader, set_up, tear_down, 0, NULL) {
+		.to_start = ut_to_start_op,
+		.ht_create = ut_ht_create_op,
+		.work_queue = ut_work_queue_op,
+		.sender_send = ut_sender_send_op,
+	};
+
+	struct leader leader = {
+		.ops = &ops,
+
+		.sigs_more = false,
+		.pages_more = false,
+		.sigs_calculated = false,
+	};
+
+	sm_init(&leader.sm, leader_sm_invariant,
+		NULL, leader_sm_conf, "leader", LS_F_ONLINE);
+
+	PRE(sm_state(&leader.sm) == LS_F_ONLINE);
+	ut_leader_message_received(&leader, append_entries());
+
+	PRE(sm_state(&leader.sm) == LS_HT_WAIT);
+	ut_disk_io(&leader.work);
+	ut_disk_io_done(&leader.work);
+
+	PRE(sm_state(&leader.sm) == LS_F_NEEDS_SNAP);
+	ut_rpc_sent(&leader.rpc);
+	ut_rpc_to_expired(&leader.rpc);
+
+	PRE(sm_state(&leader.sm) == LS_F_NEEDS_SNAP);
+	ut_rpc_sent(&leader.rpc);
+	ut_leader_message_received(&leader, ut_install_snapshot_result());
+
+	PRE(sm_state(&leader.sm) == LS_CHECK_F_HAS_SIGS);
+	ut_rpc_sent(&leader.rpc);
+	ut_leader_message_received(&leader, ut_sign_result());
+	ut_to_expired(&leader);
+
+	PRE(sm_state(&leader.sm) == LS_CHECK_F_HAS_SIGS);
+	ut_rpc_sent(&leader.rpc);
+	ut_rpc_to_expired(&leader.rpc);
+
+	PRE(sm_state(&leader.sm) == LS_CHECK_F_HAS_SIGS);
+	leader.sigs_calculated = true;
+	ut_rpc_sent(&leader.rpc);
+	ut_leader_message_received(&leader, ut_sign_result());
+
+	PRE(sm_state(&leader.sm) == LS_REQ_SIG_LOOP);
+	ut_rpc_sent(&leader.rpc);
+	PRE(sm_state(&leader.sm) == LS_REQ_SIG_LOOP);
+	ut_leader_message_received(&leader, ut_sign_result());
+	ut_disk_io(&leader.work);
+	ut_disk_io_done(&leader.work);
+	ut_disk_io(&leader.work);
+	ut_disk_io_done(&leader.work);
+
+	PRE(sm_state(&leader.sm) == LS_PAGE_READ);
+	ut_rpc_sent(&leader.rpc);
+	ut_rpc_to_expired(&leader.rpc);
+
+	PRE(sm_state(&leader.sm) == LS_PAGE_READ);
+	ut_rpc_sent(&leader.rpc);
+	ut_leader_message_received(&leader, ut_page_result());
+
+	PRE(sm_state(&leader.sm) == LS_SNAP_DONE);
+	ut_rpc_sent(&leader.rpc);
+	ut_leader_message_received(&leader, ut_install_snapshot_result());
+
+	sm_fini(&leader.sm);
+	return MUNIT_OK;
+}
+
+static void progress(void) {
+	for (unsigned i = 0; i < 100; i++) {
+		uv_run(uv_default_loop(), UV_RUN_NOWAIT);
+	}
+}
+
+static void pool_to_start_op(struct timeout *to, unsigned delay, to_cb_op cb)
+{
+	uv_timer_start(&to->handle, cb, delay, 0);
+	to->cb = cb;
+}
+
+static void pool_to_stop_op(struct timeout *to)
+{
+	uv_timer_stop(&to->handle);
+}
+
+static void pool_to_init_op(struct timeout *to)
+{
+	uv_timer_init(uv_default_loop(), &to->handle);
+}
+
+static void pool_to_expired(struct leader *leader)
+{
+	uv_timer_start(&leader->timeout.handle, leader->timeout.cb, 0, 0);
+	progress();
+}
+
+static void pool_rpc_to_expired(struct rpc *rpc)
+{
+	uv_timer_start(&rpc->timeout.handle, rpc->timeout.cb, 0, 0);
+	progress();
+}
+
+/* TODO(alberto): combine them with tests above once the rest is in place.
+ * Dispatch to one of two implementations ut or pool in general functions. */
+TEST(snapshot_leader, pool_timeouts, set_up, tear_down, 0, NULL) {
```

Comment:
[Comment for reviewer]: The idea is to keep updating this test with the pool and the rest of the implementation. I plan to keep the unit test above around as it is a very simple way of debugging the logic of the state machine.